File: ada/ada2012/Java/SymbolTable.java
import java.util.*;

public class SymbolTable {
    private Deque<Symbol> scopeStack = new ArrayDeque<>();
    private int blockCounter = 0;

    public SymbolTable() {
        Symbol globalScope = new Symbol();
        globalScope.setName("global");
        globalScope.setClassification(new HashSet<>(Arrays.asList(TypeClassification.Global_)));
        scopeStack.push(globalScope);

        defineType("integer", false);
        defineType("natural", false);
        defineType("positive", false);
        defineType("float", false);
        defineType("long_float", false);
        defineType("character", false);
        defineType("wide_character", false);
        defineType("wide_wide_character", false);
        defineType("boolean", false);
        defineType("duration", false);

        defineType("string", true);
        defineType("wide_string", true);
        defineType("wide_wide_string", true);

        definePredefined("constraint_error", TypeClassification.ExceptionName_);
        definePredefined("program_error", TypeClassification.ExceptionName_);
        definePredefined("storage_error", TypeClassification.ExceptionName_);
        definePredefined("tasking_error", TypeClassification.ExceptionName_);
        definePredefined("numeric_error", TypeClassification.ExceptionName_);

        definePredefined("true", TypeClassification.EnumerationLiteral_);
        definePredefined("false", TypeClassification.EnumerationLiteral_);
    }

    private void defineType(String name, boolean isComposite) {
        Symbol sym = new Symbol();
        sym.setName(name);
        sym.setClassification(new HashSet<>(Arrays.asList(TypeClassification.TypeName_)));
        sym.setPredefined(true);
        sym.setComposite(isComposite);
        define(sym);
    }

    private void definePredefined(String name, TypeClassification tc) {
        Symbol sym = new Symbol();
        sym.setName(name);
        sym.setClassification(new HashSet<>(Arrays.asList(tc)));
        sym.setPredefined(true);
        define(sym);
    }

    public void enterScope(Symbol newScope) {
        Symbol current = scopeStack.peek();
        if (newScope == current) return;
        scopeStack.push(newScope);
    }

    public void exitScope() {
        scopeStack.pop();
        if (scopeStack.isEmpty())
            throw new RuntimeException("SymbolTable: scope stack underflow");
    }

    public Symbol currentScope() {
        if (scopeStack.isEmpty()) return null;
        return scopeStack.peek();
    }

    public boolean define(Symbol symbol) {
        Symbol currentScope = currentScope();
        return defineInScope(currentScope, symbol);
    }

    public boolean defineInScope(Symbol currentScope, Symbol symbol) {
        symbol.setName(symbol.getName().toLowerCase());
        if (currentScope.getMembers().containsKey(symbol.getName())) {
            return false;
        }
        symbol.setParent(currentScope);
        currentScope.getMembers().put(symbol.getName(), symbol);
        return true;
    }

    public Symbol resolve(String name) {
        return resolve(name, null);
    }

    public Symbol resolve(String name, Symbol startScope) {
        name = name.toLowerCase();
        if (startScope == null) {
            for (Symbol scope : scopeStack) {
                Symbol sym = scope.getMembers().get(name);
                if (sym != null) return sym;
            }
            return null;
        } else {
            return startScope.getMembers().get(name);
        }
    }

    public Symbol pushBlockScope() {
        Symbol blockScope = new Symbol();
        blockScope.setName("block" + (++blockCounter));
        blockScope.setClassification(new HashSet<>(Arrays.asList(TypeClassification.Block_)));
        blockScope.setPredefined(true);
        enterScope(blockScope);
        return blockScope;
    }

    public void popBlockScope() {
        exitScope();
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        Symbol[] scopes = scopeStack.toArray(new Symbol[0]);
        if (scopes.length > 0) {
            toStringHelper(sb, scopes[scopes.length - 1], 0);
        }
        return sb.toString();
    }

    private void toStringHelper(StringBuilder sb, Symbol scope, int depth) {
        String indent = " ".repeat(depth * 2);
        for (Map.Entry<String, Symbol> entry : scope.getMembers().entrySet()) {
            Symbol sym = entry.getValue();
            if (!sym.isPredefined()) {
                sb.append(indent).append(sym.toString()).append("\n");
            }
            if (sym.getClassification().contains(TypeClassification.Block_) ||
                sym.getClassification().contains(TypeClassification.SubprogramName_) ||
                sym.getClassification().contains(TypeClassification.PackageName_)) {
                toStringHelper(sb, sym, depth + 1);
            }
        }
    }
}

File: ada/ada2012/Antlr4ng/SymbolTable.ts
import { Symbol } from './Symbol.js';
import { TypeClassification } from './TypeClassification.js';

export class SymbolTable {
    private scopeStack: Symbol[] = [];
    private blockCounter: number = 0;

    constructor() {
        const globalScope = new Symbol("global", new Set([TypeClassification.Global_]));
        this.scopeStack.push(globalScope);

        // Predefined scalar types
        for (const name of ["integer", "natural", "positive", "float", "long_float",
                            "character", "wide_character", "wide_wide_character", "boolean", "duration"]) {
            this.defineType(name, false);
        }
        // Predefined composite types
        for (const name of ["string", "wide_string", "wide_wide_string"]) {
            this.defineType(name, true);
        }
        // Predefined exceptions
        for (const name of ["constraint_error", "program_error", "storage_error", "tasking_error", "numeric_error"]) {
            this.definePredefined(name, TypeClassification.ExceptionName_);
        }
        // Predefined enumeration literals
        this.definePredefined("true", TypeClassification.EnumerationLiteral_);
        this.definePredefined("false", TypeClassification.EnumerationLiteral_);
    }

    private defineType(name: string, isComposite: boolean): void {
        const s = new Symbol(name, new Set([TypeClassification.TypeName_]));
        s.predefined = true;
        s.isComposite = isComposite;
        this.define(s);
    }

    private definePredefined(name: string, tc: TypeClassification): void {
        const s = new Symbol(name, new Set([tc]));
        s.predefined = true;
        this.define(s);
    }

    currentScope(): Symbol | null {
        if (this.scopeStack.length === 0) return null;
        return this.scopeStack[this.scopeStack.length - 1];
    }

    define(symbol: Symbol): boolean {
        const current = this.currentScope();
        if (current === null) return false;
        symbol.name = symbol.name.toLowerCase();
        if (current.members.has(symbol.name)) return false;
        symbol.parent = current;
        current.members.set(symbol.name, symbol);
        return true;
    }

    resolve(name: string): Symbol | null {
        const lower = name.toLowerCase();
        for (let i = this.scopeStack.length - 1; i >= 0; i--) {
            const sym = this.scopeStack[i].members.get(lower);
            if (sym !== undefined) return sym;
        }
        return null;
    }

    pushBlockScope(): Symbol {
        this.blockCounter++;
        const blockScope = new Symbol("block" + this.blockCounter, new Set([TypeClassification.Block_]));
        blockScope.predefined = true;
        this.scopeStack.push(blockScope);
        return blockScope;
    }

    popBlockScope(): void {
        if (this.scopeStack.length <= 1) throw new Error("SymbolTable: scope stack underflow");
        this.scopeStack.pop();
    }

    getExportedSymbols(): Symbol[] {
        const result: Symbol[] = [];
        if (this.scopeStack.length === 0) return result;
        const global = this.scopeStack[0];
        for (const [, sym] of global.members) {
            if (!sym.predefined) {
                result.push(sym);
            }
        }
        return result;
    }

    toString(): string {
        let sb = "";
        if (this.scopeStack.length > 0) {
            sb = this.toStringHelper(sb, this.scopeStack[0], 0);
        }
        return sb;
    }

    private toStringHelper(sb: string, scope: Symbol, depth: number): string {
        const indent = "  ".repeat(depth);
        for (const [, sym] of scope.members) {
            if (!sym.predefined) {
                sb += indent + sym.toString() + "\n";
            }
            if (sym.classification.has(TypeClassification.Block_) ||
                sym.classification.has(TypeClassification.SubprogramName_) ||
                sym.classification.has(TypeClassification.PackageName_)) {
                sb = this.toStringHelper(sb, sym, depth + 1);
            }
        }
        return sb;
    }
}
